// @aws-cdk/aws-cognito/test/test.user-pool.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const lambda = require("@aws-cdk/aws-lambda");
const cdk = require("@aws-cdk/core");
const cognito = require("../lib");
module.exports = {
    'default setup'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new cognito.UserPool(stack, 'Pool', {
            userPoolName: 'myPool',
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Cognito::UserPool', {
            UserPoolName: 'myPool'
        }));
        test.done();
    },
    'support tags'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const pool = new cognito.UserPool(stack, 'Pool', {
            userPoolName: 'myPool',
        });
        cdk.Tag.add(pool, "PoolTag", "PoolParty");
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Cognito::UserPool', {
            UserPoolName: 'myPool',
            UserPoolTags: {
                PoolTag: "PoolParty",
            }
        }));
        test.done();
    },
    'lambda triggers are defined'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const fn = new lambda.Function(stack, 'MyLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        // WHEN
        const pool = new cognito.UserPool(stack, 'Pool', {
            lambdaTriggers: {
                preSignUp: fn
            }
        });
        pool.addCustomMessageTrigger(fn);
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Cognito::UserPool', {
            LambdaConfig: {
                PreSignUp: stack.resolve(fn.functionArn),
                CustomMessage: stack.resolve(fn.functionArn)
            }
        }));
        test.done();
    },
    'on* API correctly appends triggers'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const createAuthChallengeLambdaFn = new lambda.Function(stack, 'createAuthChallengeLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const customMessageLambdaFn = new lambda.Function(stack, 'customMessageLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const defineAuthChallengeLambdaFn = new lambda.Function(stack, 'defineAuthChallengeLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const postAuthenticationLambdaFn = new lambda.Function(stack, 'postAuthenticationLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const postConfirmationLambdaFn = new lambda.Function(stack, 'postConfirmationLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const preAuthenticationLambdaFn = new lambda.Function(stack, 'preAuthenticationLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const preSignUpLambdaFn = new lambda.Function(stack, 'preSignUpLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const preTokenGenerationLambdaFn = new lambda.Function(stack, 'preTokenGenerationLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const userMigrationLambdaFn = new lambda.Function(stack, 'userMigrationLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const verifyAuthChallengeResponseLambdaFn = new lambda.Function(stack, 'verifyAuthChallengeResponseLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        // WHEN
        const pool = new cognito.UserPool(stack, 'Pool', {});
        pool.addCreateAuthChallengeTrigger(createAuthChallengeLambdaFn);
        pool.addCustomMessageTrigger(customMessageLambdaFn);
        pool.addDefineAuthChallengeTrigger(defineAuthChallengeLambdaFn);
        pool.addPostAuthenticationTrigger(postAuthenticationLambdaFn);
        pool.addPostConfirmationTrigger(postConfirmationLambdaFn);
        pool.addPreAuthenticationTrigger(preAuthenticationLambdaFn);
        pool.addPreSignUpTrigger(preSignUpLambdaFn);
        pool.addPreTokenGenerationTrigger(preTokenGenerationLambdaFn);
        pool.addUserMigrationTrigger(userMigrationLambdaFn);
        pool.addVerifyAuthChallengeResponseTrigger(verifyAuthChallengeResponseLambdaFn);
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Cognito::UserPool', {
            LambdaConfig: {
                CreateAuthChallenge: stack.resolve(createAuthChallengeLambdaFn.functionArn),
                CustomMessage: stack.resolve(customMessageLambdaFn.functionArn),
                DefineAuthChallenge: stack.resolve(defineAuthChallengeLambdaFn.functionArn),
                PostAuthentication: stack.resolve(postAuthenticationLambdaFn.functionArn),
                PostConfirmation: stack.resolve(postConfirmationLambdaFn.functionArn),
                PreAuthentication: stack.resolve(preAuthenticationLambdaFn.functionArn),
                PreSignUp: stack.resolve(preSignUpLambdaFn.functionArn),
                PreTokenGeneration: stack.resolve(preTokenGenerationLambdaFn.functionArn),
                UserMigration: stack.resolve(userMigrationLambdaFn.functionArn),
                VerifyAuthChallengeResponse: stack.resolve(verifyAuthChallengeResponseLambdaFn.functionArn)
            }
        }));
        test.done();
    },
    'lambdas are given cognito service grant'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const fn = new lambda.Function(stack, 'MyLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        // WHEN
        new cognito.UserPool(stack, 'Pool', {
            lambdaTriggers: {
                preSignUp: fn
            }
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Lambda::Permission', {
            FunctionName: stack.resolve(fn.functionArn),
            Principal: 'cognito-idp.amazonaws.com'
        }));
        test.done();
    },
    'set sign in type'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        new cognito.UserPool(stack, 'Pool', {
            signInType: cognito.SignInType.EMAIL,
            autoVerifiedAttributes: [cognito.UserPoolAttribute.EMAIL]
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::Cognito::UserPool', {
            UsernameAttributes: ['email'],
            AutoVerifiedAttributes: ['email']
        }));
        test.done();
    },
    'usernameAliasAttributes require signInType of USERNAME'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const toThrow = () => {
            new cognito.UserPool(stack, 'Pool', {
                signInType: cognito.SignInType.EMAIL,
                usernameAliasAttributes: [cognito.UserPoolAttribute.PREFERRED_USERNAME]
            });
        };
        // THEN
        test.throws(() => toThrow(), /'usernameAliasAttributes' can only be set with a signInType of 'USERNAME'/);
        test.done();
    },
    'usernameAliasAttributes must be one or more of EMAIL, PHONE_NUMBER, or PREFERRED_USERNAME'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const toThrow = () => {
            new cognito.UserPool(stack, 'Pool', {
                signInType: cognito.SignInType.USERNAME,
                usernameAliasAttributes: [cognito.UserPoolAttribute.GIVEN_NAME]
            });
        };
        // THEN
        test.throws(() => toThrow(), /'usernameAliasAttributes' can only include EMAIL, PHONE_NUMBER, or PREFERRED_USERNAME/);
        test.done();
    },
    'autoVerifiedAttributes must be one or more of EMAIL or PHONE_NUMBER'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const toThrow = () => {
            new cognito.UserPool(stack, 'Pool', {
                signInType: cognito.SignInType.EMAIL,
                autoVerifiedAttributes: [cognito.UserPoolAttribute.EMAIL, cognito.UserPoolAttribute.GENDER]
            });
        };
        // THEN
        test.throws(() => toThrow(), /'autoVerifiedAttributes' can only include EMAIL or PHONE_NUMBER/);
        test.done();
    }
};
