// @aws-cdk/aws-iam/lib/imported-resource-principal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A principal associated with an imported resource
 *
 * Some resources have roles associated with them which they assume, such as
 * Lambda Functions, CodeBuild projects, StepFunctions machines, etc.
 *
 * When those resources are imported, their actual roles are not always
 * imported with them. When that happens, we use an instance of this class
 * instead, which will add user warnings when statements are attempted to be
 * added to it.
 */
class ImportedResourcePrincipal {
    constructor(props) {
        this.assumeRoleAction = 'sts:AssumeRole';
        this.resource = props.resource;
        this.grantPrincipal = this;
    }
    get policyFragment() {
        throw new Error(`Cannot get policy fragment of ${this.resource.node.path}, resource imported without a role`);
    }
    addToPolicy(statement) {
        const repr = JSON.stringify(this.resource.node.resolve(statement));
        this.resource.node.addWarning(`Add statement to this resource's role: ${repr}`);
        return true; // Pretend we did the work. The human will do it for us, eventually.
    }
}
exports.ImportedResourcePrincipal = ImportedResourcePrincipal;
