// aws-cdk/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const colors = require("colors/safe");
const util = require("util");
// tslint:disable:no-console the whole point of those methods is precisely to output to the console...
exports.isVerbose = false;
function setVerbose(enabled = true) {
    exports.isVerbose = enabled;
}
exports.setVerbose = setVerbose;
function error(fmt, ...args) {
    console.error(colors.red(util.format(fmt, ...args)));
}
exports.error = error;
function debug(fmt, ...args) {
    if (exports.isVerbose) {
        console.error(colors.gray(util.format(fmt, ...args)));
    }
}
exports.debug = debug;
function highlight(fmt, ...args) {
    console.error(colors.bold(colors.white(util.format(fmt, ...args))));
}
exports.highlight = highlight;
function success(fmt, ...args) {
    console.error(colors.green(util.format(fmt, ...args)));
}
exports.success = success;
function warning(fmt, ...args) {
    console.error(colors.yellow(util.format(fmt, ...args)));
}
exports.warning = warning;
function print(fmt, ...args) {
    console.error(colors.white(util.format(fmt, ...args)));
}
exports.print = print;
function data(fmt, ...args) {
    console.log(util.format(fmt, ...args));
}
exports.data = data;
/**
 * Create a logger output that features a constant prefix string.
 *
 * @param prefixString the prefix string to be appended before any log entry.
 * @param fn   the logger function to be used (typically one of the other functions in this module)
 *
 * @returns a new LoggerFunction.
 */
function prefix(prefixString, fn) {
    return (fmt, ...args) => fn(`%s ${fmt}`, prefixString, ...args);
}
exports.prefix = prefix;
