// @aws-cdk/aws-cognito/lib/user-pool-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@aws-cdk/core");
const cognito_generated_1 = require("./cognito.generated");
/**
 * Types of authentication flow
 */
var AuthFlow;
(function (AuthFlow) {
    /**
     * Enable flow for server-side or admin authentication (no client app)
     */
    AuthFlow["ADMIN_NO_SRP"] = "ADMIN_NO_SRP_AUTH";
    /**
     * Enable custom authentication flow
     */
    AuthFlow["CUSTOM_FLOW_ONLY"] = "CUSTOM_AUTH_FLOW_ONLY";
    /**
     * Enable auth using username & password
     */
    AuthFlow["USER_PASSWORD"] = "USER_PASSWORD_AUTH";
})(AuthFlow = exports.AuthFlow || (exports.AuthFlow = {}));
/**
 * Define a UserPool App Client
 */
class UserPoolClient extends core_1.Resource {
    constructor(scope, id, props) {
        super(scope, id, {
            physicalName: props.userPoolClientName,
        });
        const resource = new cognito_generated_1.CfnUserPoolClient(this, 'Resource', {
            clientName: this.physicalName,
            generateSecret: props.generateSecret,
            userPoolId: props.userPool.userPoolId,
            explicitAuthFlows: props.enabledAuthFlows
        });
        this.userPoolClientId = resource.ref;
        this.userPoolClientClientSecret = resource.attrClientSecret;
        this.userPoolClientName = resource.attrName;
    }
}
exports.UserPoolClient = UserPoolClient;
