// aws-cdk/test/api/test.deploy-stack.js
"use strict";
const lib_1 = require("../../lib");
const mock_sdk_1 = require("../util/mock-sdk");
const FAKE_STACK = {
    name: 'withouterrors',
    template: { resource: 'noerrorresource' },
    environment: { name: 'dev', account: '12345', region: 'here' },
    metadata: {},
};
module.exports = {
    async 'do deploy executable change set with 0 changes'(test) {
        // GIVEN
        const sdk = new mock_sdk_1.MockSDK();
        let executed = false;
        sdk.stubCloudFormation({
            describeStacks() {
                return {
                    Stacks: []
                };
            },
            createChangeSet() {
                return {};
            },
            describeChangeSet() {
                return {
                    Status: 'CREATE_COMPLETE',
                    Changes: [],
                };
            },
            executeChangeSet() {
                executed = true;
                return {};
            }
        });
        // WHEN
        const ret = await lib_1.deployStack({
            stack: FAKE_STACK,
            sdk,
        });
        // THEN
        test.equals(ret.noOp, false);
        test.equals(executed, true);
        test.done();
    },
};
