// @aws-cdk/aws-lambda-event-sources/lib/dynamodb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dynamodb = require("@aws-cdk/aws-dynamodb");
const lambda = require("@aws-cdk/aws-lambda");
/**
 * Use an Amazon DynamoDB stream as an event source for AWS Lambda.
 */
class DynamoEventSource {
    constructor(table, props) {
        this.table = table;
        this.props = props;
        if (this.props.batchSize !== undefined && (this.props.batchSize < 1 || this.props.batchSize > 1000)) {
            throw new Error(`Maximum batch size must be between 1 and 1000 inclusive (given ${this.props.batchSize})`);
        }
    }
    bind(target) {
        new lambda.EventSourceMapping(target, `DynamoDBEventSource:${this.table.node.uniqueId}`, {
            target,
            batchSize: this.props.batchSize || 100,
            eventSourceArn: this.table.tableStreamArn,
            startingPosition: this.props.startingPosition
        });
        this.table.grantStreamRead(target);
        dynamodb.Table.grantListStreams(target);
    }
}
exports.DynamoEventSource = DynamoEventSource;
