// @aws-cdk/cdk/lib/cloudformation-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const instrinsics_1 = require("./instrinsics");
const resolve_1 = require("./resolve");
const token_1 = require("./token");
/**
 * Class for JSON routines that are framework-aware
 */
class CloudFormationJSON {
    /**
     * Turn an arbitrary structure potentially containing Tokens into a JSON string.
     *
     * Returns a Token which will evaluate to CloudFormation expression that
     * will be evaluated by CloudFormation to the JSON representation of the
     * input structure.
     *
     * All Tokens substituted in this way must return strings, or the evaluation
     * in CloudFormation will fail.
     *
     * @param obj The object to stringify
     * @param context The Construct from which to resolve any Tokens found in the object
     */
    static stringify(obj, context) {
        return new token_1.Token(() => {
            // Resolve inner value first so that if they evaluate to literals, we
            // maintain the type (and discard 'undefined's).
            //
            // Then replace intrinsics with a special subclass of Token that
            // overrides toJSON() to the marker string, so if we resolve() the
            // strings again it evaluates to the right string. It also
            // deep-escapes any strings inside the intrinsic, so that if literal
            // strings are used in {Fn::Join} or something, they will end up
            // escaped in the final JSON output.
            const resolved = resolve_1.resolve(obj, {
                scope: context,
                prefix: []
            });
            // We can just directly return this value, since resolve() will be called
            // on our return value anyway.
            return JSON.stringify(deepReplaceIntrinsics(resolved));
        }).toString();
        /**
         * Recurse into a structure, replace all intrinsics with IntrinsicTokens.
         */
        function deepReplaceIntrinsics(x) {
            if (x == null) {
                return x;
            }
            if (instrinsics_1.isIntrinsic(x)) {
                return wrapIntrinsic(x);
            }
            if (Array.isArray(x)) {
                return x.map(deepReplaceIntrinsics);
            }
            if (typeof x === 'object') {
                for (const key of Object.keys(x)) {
                    x[key] = deepReplaceIntrinsics(x[key]);
                }
            }
            return x;
        }
        function wrapIntrinsic(intrinsic) {
            return new IntrinsicToken(() => deepQuoteStringsForJSON(intrinsic));
        }
    }
}
exports.CloudFormationJSON = CloudFormationJSON;
/**
 * Token that also stringifies in the toJSON() operation.
 */
class IntrinsicToken extends token_1.Token {
    /**
     * Special handler that gets called when JSON.stringify() is used.
     */
    toJSON() {
        return this.toString();
    }
}
/**
 * Deep escape strings for use in a JSON context
 */
function deepQuoteStringsForJSON(x) {
    if (typeof x === 'string') {
        // Whenever we escape a string we strip off the outermost quotes
        // since we're already in a quoted context.
        const stringified = JSON.stringify(x);
        return stringified.substring(1, stringified.length - 1);
    }
    if (Array.isArray(x)) {
        return x.map(deepQuoteStringsForJSON);
    }
    if (typeof x === 'object') {
        for (const key of Object.keys(x)) {
            x[key] = deepQuoteStringsForJSON(x[key]);
        }
    }
    return x;
}
