// @aws-cdk/cdk/lib/pseudo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cfn_reference_1 = require("./cfn-reference");
const token_1 = require("./token");
const AWS_ACCOUNTID = 'AWS::AccountId';
const AWS_URLSUFFIX = 'AWS::URLSuffix';
const AWS_NOTIFICATIONARNS = 'AWS::NotificationARNs';
const AWS_PARTITION = 'AWS::Partition';
const AWS_REGION = 'AWS::Region';
const AWS_STACKID = 'AWS::StackId';
const AWS_STACKNAME = 'AWS::StackName';
const AWS_NOVALUE = 'AWS::NoValue';
/**
 * Accessor for pseudo parameters
 *
 * Since pseudo parameters need to be anchored to a stack somewhere in the
 * construct tree, this class takes an scope parameter; the pseudo parameter
 * values can be obtained as properties from an scoped object.
 */
class Aws {
    constructor() {
    }
    static get accountId() {
        return new UnscopedPseudo(AWS_ACCOUNTID).toString();
    }
    static get urlSuffix() {
        return new UnscopedPseudo(AWS_URLSUFFIX).toString();
    }
    static get notificationArns() {
        return new UnscopedPseudo(AWS_NOTIFICATIONARNS).toList();
    }
    static get partition() {
        return new UnscopedPseudo(AWS_PARTITION).toString();
    }
    static get region() {
        return new UnscopedPseudo(AWS_REGION).toString();
    }
    static get stackId() {
        return new UnscopedPseudo(AWS_STACKID).toString();
    }
    static get stackName() {
        return new UnscopedPseudo(AWS_STACKNAME).toString();
    }
    static get noValue() {
        return new UnscopedPseudo(AWS_NOVALUE).toString();
    }
}
exports.Aws = Aws;
/**
 * Accessor for scoped pseudo parameters
 *
 * These pseudo parameters are anchored to a stack somewhere in the construct
 * tree, and their values will be exported automatically.
 */
class ScopedAws {
    constructor(scope) {
        this.scope = scope;
    }
    get accountId() {
        return new ScopedPseudo(AWS_ACCOUNTID, this.scope).toString();
    }
    get urlSuffix() {
        return new ScopedPseudo(AWS_URLSUFFIX, this.scope).toString();
    }
    get notificationArns() {
        return new ScopedPseudo(AWS_NOTIFICATIONARNS, this.scope).toList();
    }
    get partition() {
        return new ScopedPseudo(AWS_PARTITION, this.scope).toString();
    }
    get region() {
        return new ScopedPseudo(AWS_REGION, this.scope).toString();
    }
    get stackId() {
        return new ScopedPseudo(AWS_STACKID, this.scope).toString();
    }
    get stackName() {
        return new ScopedPseudo(AWS_STACKNAME, this.scope).toString();
    }
}
exports.ScopedAws = ScopedAws;
class ScopedPseudo extends cfn_reference_1.CfnReference {
    constructor(name, scope) {
        super({ Ref: name }, name, scope);
    }
}
class UnscopedPseudo extends token_1.Token {
    constructor(name) {
        super({ Ref: name }, name);
    }
}
