// @aws-cdk/cx-api/lib/cxapi.js
"use strict";
/**
 * File with definitions for the interface between the Cloud Executable and the CDK toolkit.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Bump this to the library version if and only if the CX protocol changes.
 *
 * We could also have used 1, 2, 3, ... here to indicate protocol versions, but
 * those then still need to be mapped to software versions to be useful. So we
 * might as well use the software version as protocol version and immediately
 * generate a useful error message from this.
 *
 * Note the following:
 *
 * - The versions are not compared in a semver way, they are used as
 *    opaque ordered tokens.
 * - The version needs to be set to the NEXT releasable version when it's
 *   updated (as the current verison in package.json has already been released!)
 * - The request does not have versioning yet, only the response.
 */
exports.PROTO_RESPONSE_VERSION = '0.19.0';
/**
 * The name of the root manifest file of the assembly.
 */
exports.MANIFEST_FILE = 'manifest.json';
/**
 * The name of the root file with build instructions.
 */
exports.BUILD_FILE = 'build.json';
exports.OUTDIR_ENV = 'CDK_OUTDIR';
exports.CONTEXT_ENV = 'CDK_CONTEXT_JSON';
/**
 * Context parameter for the default AWS account to use if a stack's environment is not set.
 */
exports.DEFAULT_ACCOUNT_CONTEXT_KEY = 'aws:cdk:toolkit:default-account';
/**
 * Context parameter for the default AWS region to use if a stack's environment is not set.
 */
exports.DEFAULT_REGION_CONTEXT_KEY = 'aws:cdk:toolkit:default-region';
/**
 * Metadata key used to print INFO-level messages by the toolkit when an app is syntheized.
 */
exports.INFO_METADATA_KEY = 'aws:cdk:info';
/**
 * Metadata key used to print WARNING-level messages by the toolkit when an app is syntheized.
 */
exports.WARNING_METADATA_KEY = 'aws:cdk:warning';
/**
 * Metadata key used to print ERROR-level messages by the toolkit when an app is syntheized.
 */
exports.ERROR_METADATA_KEY = 'aws:cdk:error';
/**
 * The key used when CDK path is embedded in **CloudFormation template**
 * metadata.
 */
exports.PATH_METADATA_KEY = 'aws:cdk:path';
/**
 * Enables the embedding of the "aws:cdk:path" in CloudFormation template metadata.
 */
exports.PATH_METADATA_ENABLE_CONTEXT = 'aws:cdk:enable-path-metadata';
/**
 * Disables the emission of `cdk.out`
 */
exports.DISABLE_LEGACY_MANIFEST_CONTEXT = 'aws:cdk:disable-legacy-manifest';
/**
 * The name of the pre 0.25.0 manifest file. Will only be emitted if
 * aws:cdk:disable-legacy-manifest is not defined.
 *
 * @deprecated Use `MANIFEST_FILE`
 */
exports.OUTFILE_NAME = 'cdk.out';
/**
 * Disable the collection and reporting of version information.
 */
exports.DISABLE_VERSION_REPORTING = 'aws:cdk:disable-version-reporting';
