// @aws-cdk/aws-kms/test/test.via-service-principal.js
"use strict";
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/cdk");
const kms = require("../lib");
module.exports = {
    'Via service, any principal'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const statement = new iam.PolicyStatement()
            .addAction('abc:call')
            .addPrincipal(new kms.ViaServicePrincipal('bla.amazonaws.com'))
            .addResource('*');
        // THEN
        test.deepEqual(stack.node.resolve(statement), {
            Action: 'abc:call',
            Condition: { StringEquals: { 'kms:ViaService': 'bla.amazonaws.com' } },
            Effect: 'Allow',
            Principal: '*',
            Resource: '*'
        });
        test.done();
    },
    'Via service, principal with conditions'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const statement = new iam.PolicyStatement()
            .addAction('abc:call')
            .addPrincipal(new kms.ViaServicePrincipal('bla.amazonaws.com', new iam.OrganizationPrincipal('o-1234')))
            .addResource('*');
        // THEN
        test.deepEqual(stack.node.resolve(statement), {
            Action: 'abc:call',
            Condition: {
                StringEquals: {
                    'kms:ViaService': 'bla.amazonaws.com',
                    'aws:PrincipalOrgID': 'o-1234'
                }
            },
            Effect: 'Allow',
            Principal: '*',
            Resource: '*'
        });
        test.done();
    },
};
