// @aws-cdk/cdk/test/test.parameter.js
"use strict";
const lib_1 = require("../lib");
module.exports = {
    'parameters can be used and referenced using param.ref'(test) {
        const stack = new lib_1.Stack();
        const child = new lib_1.Construct(stack, 'Child');
        const param = new lib_1.CfnParameter(child, 'MyParam', {
            default: 10,
            type: 'Integer',
            description: 'My first parameter'
        });
        new lib_1.CfnResource(stack, 'Resource', { type: 'Type', properties: { ReferenceToParam: param.ref } });
        test.deepEqual(stack._toCloudFormation(), {
            Parameters: {
                ChildMyParam3161BF5D: {
                    Default: 10,
                    Type: 'Integer',
                    Description: 'My first parameter'
                }
            },
            Resources: {
                Resource: {
                    Type: 'Type',
                    Properties: { ReferenceToParam: { Ref: 'ChildMyParam3161BF5D' } }
                }
            }
        });
        test.done();
    },
    'parameters are tokens, so they can be assigned without .ref and their Ref will be taken'(test) {
        const stack = new lib_1.Stack();
        const param = new lib_1.CfnParameter(stack, 'MyParam', { type: 'String' });
        test.deepEqual(stack.node.resolve(param), { Ref: 'MyParam' });
        test.done();
    }
};
